Added support for querying version information in the Qiskit C API.

velocityraptor7085 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

- Updated the Qiskit C extension to enable querying the Qiskit major, minor and patch version information.
- This fixes the following open issue: https://github.com/Qiskit/qiskit/issues/14280


### Details and comments

The approach used includes the following changes:
1) Adds a file called `version.h` in the directory `crates/cext` having macros for the version information.
2) Includes the `version.h` macros in the `qiskit.h` file generated for the C API calls (added through the `cbindgen.toml` file within `crates/cext`).
3) The version information is populated into the `version.h` file by reading the `VERSION.txt` file (within the `qiskit` subdirectory) by the `build.rs` code via regex matching and replacement at compile time (for this an additional build dependency `regex = "1"` was added to the `Cargo.toml` file in the `cext` crate). The removes the need to update the header for every new release.

- All code was formatted via `cargo fmt` and passed the `tox -elint` test.
- The modified branch was tested and validated via the `test_version_info.c` file in the test/c directory.

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14423) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14423) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14423) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: velocityraptor7085<br/>:x: Archit Chadalawada<br/><hr/>**Archit Chadalawada** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit?pullRequest=14423) it.</sub>

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

mtreinish reviewed (commented):
Thanks for doing this, I think the approach you're using here is a bit more complicated than it needs to be. Couldn't we just define a couple of pub constants in the cext rust code and then tell cbindgen to include them in the header? If so then we can just use the cargo env variables to populate the version numbers appropriately.

mtreinish commented:
Oh also you appear to have authored commits with a different email address not associated with your github account. You'll either need to associate that email account to your github account, or edit the commits on your branch to reset the author metadata to use the correct email address and force push that updated history to your PR's branch on your fork. We won't be able to merge a PR if the CLA bot isn't all green.

velocityraptor7085 commented:
Okay, I am working on it. 👍

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

velocityraptor7085 commented:
I have updated the code as well as force pushed the correct author information.

### Summary

- Updated the Qiskit C extension to enable querying the Qiskit major, minor and patch version information.
- This fixes the following open issue: https://github.com/Qiskit/qiskit/issues/14280

### Details and comments

The approach used includes the following changes:
1) Added macros for numerical Qiskit versions in the `cbindgen.toml` file (within `crates/cext`).
2) Updated the `build.rs` file in the `crates/cext` directory to fetch version information at compile-time via the `CARGO_PKG_VERSION_...` environment variables.

This enables the `cbindgen.toml` file to directly include the version-related macros in the generated `qiskit.h` header file, thereby making it directly accessible via the Qiskit C API.

Please let me know if this approach is okay.
Thank you.

velocityraptor7085 committed:
Added support for querying version information in the Qiskit C API.

Cryoris commented:
I'm not super sure about this approach, since we effectively end up with some version number committed in `cbindgen.toml` and dynamically overwrite them. So when we update Qiskit, we either (1) end up with an out-of-date version in `cbindgen.toml` or (2) manually update it, but then we wouldn't need the dynamic overwrite in the first place.

I was hoping we could have a solution where we could dynamically add the version numbers to the header, but after scouting the cbindgen docs I couldn't find anything suitable that allows to extend e.g. the `after_includes` (only to replace it completely). 

I'm wondering whether at this point it wouldn't just be easier to hardcode the version number? Or a compromise, if we want to stick with the cargo-generated version, we could maybe have a placeholder comment in the `after_includes` of `cbindgen.toml` that doesn't hardcode version numbers. (@velocityraptor7085 nothing you have to change yet, this is to discuss with @mtreinish first 🙂)

velocityraptor7085 commented:
Sure @Cryoris . Actually, the `cbindgen.toml` was initially having a placeholder (not hardcoded values) which were defined. The idea was that whenever the `make c` command was run by the user, the up-to-date Qiskit version (as part of the environment variables) will overwrite whatever comes after the `#define ...` in the `after_includes` section of the `cbindgen.toml` (The logic for dynamically writing this is in the `build.rs` file).

So, If you try testing this by removing the version numbers from the `cbindgen.toml` and re-running `make c`, the version numbers should automatically be populated. I can update the `cbindgen.toml` with some placeholder like `@` or something else suitable to avoid any ambiguity.

Please let me know in case I am missing something.
Thank you.

Cryoris commented:
We discussed offline with @mtreinish and came to the conclusion that it's probably simpler to just hardcode the version number, without any additional cbindgen logic. That is partially motivated by #14430, which we'll likely change to sooner than later, which drops cbindgen anyways. Could you update the code accordingly? 🙂

velocityraptor7085 committed:
Updated cbindgen.toml with version numbers for the Qiskit C API

velocityraptor7085 commented:
Okay @Cryoris. I have updated the code to use the hard-coded version information directly without any additional logic related to `cbindgen.toml`. Let me know if this works. Thank you.

Cryoris commented on crates/cext/cbindgen.toml:
```diff
@@ -10,6 +10,13 @@ after_includes = """
 
+#define QISKIT_VERSION_MAJOR 2
+#define QISKIT_VERSION_MINOR 0
+#define QISKIT_VERSION_PATCH 2
```
This will be included in the 2.1.0 release 🙂 2.0.2 is just a bugfix release without any new features (like this one)
```suggestion
#define QISKIT_VERSION_MINOR 1
#define QISKIT_VERSION_PATCH 0
```

Cryoris commented on crates/cext/cbindgen.toml:
```diff
@@ -56,4 +63,4 @@ prefix_with_name = true
 "OpCount" = "QkOpCount"
 "CInstruction" = "QkCircuitInstruction"
-"ExitCode" = "QkExitCode"
+"ExitCode" = "QkExitCode"
```
Very minor nit: could you keep the final linebreaks inplace?

Cryoris commented on test/c/test_version_info.c:
```diff
@@ -0,0 +1,21 @@
+#include "common.h"
+
+// Test that QISKIT_VERSION_MAJOR, _MINOR, _PATCH are defined and print them
+int test_version_macros() {
```
Could you add a test for the `QISKIT_VERSION` and `QISKIT_VERSION_NUMERIC` macros?

Cryoris reviewed (commented):
Thanks for the updates, this generally looks good to me now. I left some small comments below.

velocityraptor7085 committed:
Update crates/cext/cbindgen.toml

Co-authored-by: Julien Gacon <gaconju@gmail.com>

velocityraptor7085 committed:
Updated cbindgen.toml with version numbers for the Qiskit C API

Cryoris commented on test/c/test_version_info.c:
```diff
@@ -0,0 +1,26 @@
+    if (QISKIT_VERSION_MAJOR < 0 || QISKIT_VERSION_MINOR < 0 || QISKIT_VERSION_PATCH < 0) {
+        return EqualityError;
+    }
+    if (QISKIT_VERSION_NUMERIC(QISKIT_VERSION_MAJOR, QISKIT_VERSION_MINOR, QISKIT_VERSION_PATCH) !=
```
Hang on, if these are the same, why do we need the `QISKIT_VERSION_NUMERIC` macro? 🤔

velocityraptor7085 replied:
An application developer may use it in his code in the following manner to check for version safeness:
if (QISKIT_VERSION >= QISKIT_VERSION_NUMERIC(2,1,0)) { ... }

velocityraptor7085 commented:
Thank you so much! 🙂 I have made the suggested changes @Cryoris.
BTW, I'm just curious as to why it's a convention to leave an empty blank line at the end of the `cbindgen.toml` as I am not very experienced with this.

Cryoris commented on crates/cext/cbindgen.toml:
```diff
@@ -9,6 +9,33 @@ after_includes = """
+#endif
+
+#endif  //__cplusplus
+
```
These don't belong here, that's probably a merge artifact 🙂 
```suggestion
```

Cryoris commented:
@velocityraptor7085 the history of this PR includes some commits that don't belong here (e.g. [beeb97c](https://github.com/Qiskit/qiskit/pull/14423/commits/beeb97cd28bda9eb77e3227e7619e26edf0efdd2)). Would you mind rebasing to remove these or just clean the history? 

> BTW, I'm just curious as to why it's a convention to leave an empty blank line at the end of the cbindgen.toml as I am not very experienced with this.

It's often considered good practice because some tools expect the last line to be a linebreak (see e.g. https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file)

Another note: you don't have to keep updating your branch with the `main` version, that's not required to merge your PR and it keeps triggering CI to run 🙂

Cryoris commented:
Also, let us know if you need help cleaning up the history.

velocityraptor7085 committed:
updated Makefile

velocityraptor7085 committed:
updated Makefile

velocityraptor7085 commented:
Thanks a lot. I just made the changes. Let me know if the changes are good to go.

mtreinish commented on test/c/test_version_info.c:
```diff
@@ -0,0 +1,26 @@
+// Test that QISKIT_VERSION_MAJOR, _MINOR, _PATCH are defined and print them
+int test_version_macros() {
+    printf("Qiskit version: %d.%d.%d\n", QISKIT_VERSION_MAJOR, QISKIT_VERSION_MINOR,
+           QISKIT_VERSION_PATCH);
```
Should we be printing this without a condition? I wouldn't expect anything to be written to stdout on success.

velocityraptor7085 replied:
Yes that makes sense, I just removed the print statement since they are tested anyways in the next line of code. Thank you.

mtreinish reviewed (commented):
Thanks for doing this, I think it mostly looks good. Just one small inline question. Besides that can you add a release note to document the addition to the C API. The documentation around this is here: https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes

velocityraptor7085 committed:
Removed printf statement in the test file

velocityraptor7085 commented:
Okay I'll have a look and add a release note 👍

velocityraptor7085 committed:
Added release note for C API version support

velocityraptor7085 commented:
@mtreinish Please let me know if the changes are appropriate. Thank you.

mtreinish commented on releasenotes/notes/c-api-version-support-55b018bf9552db1f.yaml:
````diff
@@ -0,0 +1,20 @@
+    if (QISKIT_VERSION >= QISKIT_VERSION_NUMERIC(2, 1, 0)) {
+        // Code for version 2.1.0 or later
+    }
+    ```
````
The reno release notes use sphinx and restructured text. So the markdown formatting you used here won't render correctly:

```suggestion
features_c:
  - |
    Added support for querying the Qiskit version information from the C API via the following macros:
    
    * ``QISKIT_VERSION_MAJOR`` : Contains the major version number.
    * ``QISKIT_VERSION_MINOR`` : Contains the minor version number.
    * ``QISKIT_VERSION_PATCH`` : Contains the patch version number.
    * ``QISKIT_VERSION`` : Contains a numeric representation of the version information, which can be used for comparisons.
    * ``QISKIT_VERSION_NUMERIC(M,m,p)`` : Function-like Macro that returns the version "M.m.p" as a numeric value, which can be used for comparisons.

    For example, to check if the current version is at least 2.1.0, you can use:

    .. code-block: c

        if (QISKIT_VERSION >= QISKIT_VERSION_NUMERIC(2, 1, 0)) {
            // Code for version 2.1.0 or later
        }
```

mtreinish commented on releasenotes/notes/c-api-version-support-55b018bf9552db1f.yaml:
```diff
@@ -0,0 +1,20 @@
+fixes:
+  - |
+    This fixes the issue where there is currently no way to query the Qiskit version information from the C API.
+    Refer to `#14280 <https://github.com/Qiskit/qiskit/issues/14280>` for more details.
```
You don't need a fix note here. This is just a new feature.
```suggestion
```

mtreinish reviewed (commented):
Just some comments on the release notes, thanks for adding them.

velocityraptor7085 committed:
Update releasenotes/notes/c-api-version-support-55b018bf9552db1f.yaml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

velocityraptor7085 committed:
Update releasenotes/notes/c-api-version-support-55b018bf9552db1f.yaml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

velocityraptor7085 commented:
Thanks a lot for your help @mtreinish !

mtreinish commented on test/c/test_version_info.c:
```diff
@@ -0,0 +1,24 @@
+#include "common.h"
```
This is missing a copyright and license header. This is pointing out a gap in our CI that we're not enforcing this on the C test files. 

```suggestion
// This code is part of Qiskit.
//
// (C) Copyright IBM 2025.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

#include "common.h"
```

mtreinish reviewed (commented):
This LGTM now thanks for the updates. I left a small comment inline about a missing copyright header. I'll apply that and then approve this for merge.

mtreinish committed:
Add missing license header to the test file
